File: 1038.cpp
struct Solution {
    void rot(TreeNode *node, int &cumsum){
        if (node->right != nullptr){
            this->rot(node->right, cumsum);
        }
        node->val += cumsum;
        cumsum = node->val;
        if (node->left != nullptr){
            this->rot(node->left, cumsum);
        }
    }
    inline TreeNode* bstToGst(TreeNode* root) {
        int cumsum = 0;
        this->rot(root, cumsum);
        return root;
    }
};

File: sol/1038.py
class Solution:
    def bstToGst(self, root: TreeNode) -> TreeNode:
        cumsum = 0
        def rot(node: TreeNode) -> int:
            nonlocal cumsum
            if node.right is not None:
                rot(node.right)
            node.val += cumsum
            cumsum = node.val
            if node.left is not None:
                rot(node.left)
        rot(root)
        return root

